File: Problem/By section/Prefix Sum/Problem 303/problem_303.cpp
#include <bits/stdc++.h>

using namespace std;

class NumArray {
public:
    vector<int> array;
    vector<int> dp;

    NumArray(vector<int>& nums) {
        int size = nums.size();

        array = vector<int>(nums);
        dp = vector<int>(size + 1, 0);
        int result = 0;

        for(int i = 1; i <= size; i ++) {
            result += array[i - 1];
            dp[i] = result;
        }
    }
    
    // Time: O(1), Space: O(n)
    int sumRange(int left, int right) {
        return(dp[right + 1] - dp[left]);
    }
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * int param_1 = obj->sumRange(left,right);
 */

int main(int argc, char** argv) {

    return 0;
}

File: Problem/By section/Prefix Sum/Problem 303/problem_303.java
class NumArray{
    // sums[i] = the sum from nums[0] to nums[i - 1]
    private int[] sums;

    // Time: O(n), Space:O(n)
    public NumArray(int[] nums) {
        sums = new int[nums.length + 1];
        for (int i = 1; i <= nums.length; i++) {
            sums[i] = sums[i - 1] + nums[i - 1];
        }
    }

    //sumRange(i, j) =  sum[j + 1] - sum[i]
    public int sumRange(int i, int j) {
        return sums[j + 1] - sums[i];
    }
}